### ISSUE TITLE
Predbat not working on all updates after v17.19.0 - Exception Raised 1440/1470

### ISSUE THREAD

**Issue by Neomancer86** [springfall2008/batpred#1127]
**Describe the bug**
Predbat does not start, no automation and the following errors in status
7.19.1 Changed to ERROR: Exception raised 1440
7.19.2 Changed to ERROR: Exception raised 1470

**Expected behavior**
Predbat to run as normal, i'm having to stay on an older version

**Predbat version**

17.19.1
17.19.2
17.19.3
17.19.4
17.19.5

**Environment details**

- GivEnergy + 9.5 Battery
- Standard HAOS installer 
- Anything else?

**Screenshots**
If applicable, add screenshots to help explain your problem. The most useful ones can be your battery chart, the Predbat HTML plan and your current settings in HA.

**Log file**
-----------------------------------------------------------
 Add-on: AppDaemon
 Python Apps and Dashboard using AppDaemon 4.x for Home Assistant
-----------------------------------------------------------
 Add-on version: 0.16.5
 You are running the latest version of this add-on.
 System: Home Assistant OS 12.3  (amd64 / qemux86-64)
 Home Assistant Core: 2024.5.4
 Home Assistant Supervisor: 2024.05.1
-----------------------------------------------------------
 Please, share the above information when looking for help
 or support in, e.g., GitHub, forums or the Discord chat.
-----------------------------------------------------------
s6-rc: info: service base-addon-banner successfully started
s6-rc: info: service fix-attrs: starting
s6-rc: info: service base-addon-log-level: starting
s6-rc: info: service fix-attrs successfully started
s6-rc: info: service base-addon-log-level successfully started
s6-rc: info: service legacy-cont-init: starting
s6-rc: info: service legacy-cont-init successfully started
s6-rc: info: service init-appdaemon: starting
s6-rc: info: service init-appdaemon successfully started
s6-rc: info: service appdaemon: starting
s6-rc: info: service appdaemon successfully started
s6-rc: info: service legacy-services: starting
[10:38:27] INFO: Starting AppDaemon...
s6-rc: info: service legacy-services successfully started
Update complete
2024-05-20 13:16:50.276060 WARNING pred_bat: ------------------------------------------------------------
2024-05-20 13:16:50.276359 WARNING pred_bat: Unexpected error in worker for App pred_bat:
2024-05-20 13:16:50.276621 WARNING pred_bat: Worker Ags: {'id': '9448d5d3ccb1441aab8099b46afa0bcd', 'name': 'pred_bat', 'objectid': '4bd645716c5044a69c7ffecb95ba4992', 'type': 'scheduler', 'function': <bound method PredBat.update_time_loop of <predbat.PredBat object at 0x7feb993f0510>>, 'pin_app': True, 'pin_thread': 0, 'kwargs': {'interval': 15, 'random_start': 0, 'random_end': 0, '__thread_id': 'thread-0'}}
2024-05-20 13:16:50.276687 WARNING pred_bat: ------------------------------------------------------------
2024-05-20 13:16:50.277999 WARNING pred_bat: Traceback (most recent call last):
  File "/usr/lib/python3.11/site-packages/appdaemon/threading.py", line 1022, in worker
    funcref(self.AD.sched.sanitize_timer_kwargs(app, args["kwargs"]))
  File "/usr/lib/python3.11/site-packages/appdaemon/adbase.py", line 35, in f_app_lock
    return f(*args, **kw)
           ^^^^^^^^^^^^^^
  File "/homeassistant/appdaemon/apps/batpred/predbat.py", line 15052, in update_time_loop
    raise e
  File "/homeassistant/appdaemon/apps/batpred/predbat.py", line 15047, in update_time_loop
    self.update_pred(scheduled=False)
  File "/usr/lib/python3.11/site-packages/appdaemon/adbase.py", line 35, in f_app_lock
    return f(*args, **kw)
           ^^^^^^^^^^^^^^
  File "/homeassistant/appdaemon/apps/batpred/predbat.py", line 13937, in update_pred
    recompute = self.calculate_plan(recompute=recompute)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/homeassistant/appdaemon/apps/batpred/predbat.py", line 12196, in calculate_plan
    self.calculate_yesterday()
  File "/homeassistant/appdaemon/apps/batpred/predbat.py", line 11993, in calculate_yesterday
    ) = self.run_prediction(charge_limit_best, charge_window_best, [], [], False, end_record=(24 * 60), save="yesterday")
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/homeassistant/appdaemon/apps/batpred/predbat.py", line 6203, in run_prediction
    ) = pred.run_prediction(charge_limit, charge_window, discharge_window, discharge_limits, pv10, end_record, save, step)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/homeassistant/appdaemon/apps/batpred/predbat.py", line 2090, in run_prediction
    predict_pv_power[stamp] = round((pv_forecast_minute_step[minute] + pv_forecast_minute_step.get(minute + step, 0)) * (30 / step), 3)
                                     ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^
KeyError: 1470

2024-05-20 13:16:50.278105 WARNING pred_bat: ------------------------------------------------------------



Predbat Logs:
2024-05-20 13:16:39.193481 INFO pred_bat: --------------- PredBat - update at 2024-05-20 13:15:00+01:00 with clock skew 0 minutes, minutes now 795
2024-05-20 13:16:39.568071 INFO pred_bat: Predbat /homeassistant/appdaemon/apps/batpred/predbat.py version v7.19.5 currently running, latest version is v7.19.5 latest beta v7.19.5
2024-05-20 13:16:39.568597 INFO pred_bat: Autoupdate: Currently up to date
2024-05-20 13:16:39.587676 INFO pred_bat: Configuration: forecast_hours 24 num_cars 0 debug enable is True calculate_plan_every 5
2024-05-20 13:16:39.588464 INFO pred_bat: Predbat mode is set to Control charge & discharge
2024-05-20 13:16:39.897913 INFO pred_bat: Found 21600 load_today datapoints going back 14 days
2024-05-20 13:16:40.271991 INFO pred_bat: Car charging hold False threshold 6.0
2024-05-20 13:16:40.272535 INFO pred_bat: Current data so far today: load 7.05 kWh import 1.6 kWh export 4.5 kWh pv 16.8 kWh
2024-05-20 13:16:40.273091 INFO pred_bat: Fetching futurerate data from None
2024-05-20 13:16:40.273610 INFO pred_bat: Info: Fetch Octopus rates from sensor.octopus_energy_electricity_21e5214280_2200018159199_current_rate
2024-05-20 13:16:40.283113 INFO pred_bat: Info: Fetch Octopus rates from sensor.octopus_energy_electricity_21e5214280_2200043685611_export_current_rate
2024-05-20 13:16:40.291892 INFO pred_bat: Standing charge is set to 65.121 p
2024-05-20 13:16:40.409465 INFO pred_bat: Rate min forward looking: now 14.52 at end of forecast 14.52
2024-05-20 13:16:40.410339 INFO pred_bat: Import rates min 14.52 max 33.89 average 24.21
2024-05-20 13:16:40.414534 INFO pred_bat: Export rates min 5.16 max 24.52 average 14.84
2024-05-20 13:16:40.415736 INFO pred_bat: Rate thresholds (for charge/discharge) are import 33.39p (0.0) export 5.66p (0.0)
2024-05-20 13:16:40.417440 INFO pred_bat: High export rate found rates in range 14.84 to 24.52
2024-05-20 13:16:40.418928 INFO pred_bat: Low Import rate found rates in range 14.52 to 24.21
2024-05-20 13:16:40.436042 INFO pred_bat: Todays energy import 1.6 kWh export 4.5 kWh cost 23.48 p import 90.25 p export -66.77 p carbon 0.0 kg
2024-05-20 13:16:40.437935 INFO pred_bat: WARN: No solar data has been configured.
2024-05-20 13:16:40.440131 INFO pred_bat: Historical data totals for days [7, 14] are [20.7, 14.53] - min 14.53
2024-05-20 13:16:40.441112 INFO pred_bat: Historical days now [7, 14] weight [1, 1]
2024-05-20 13:16:40.450698 INFO pred_bat: Today's load divergence 103.21 % in-day adjustment 103.21 % damping 1.0x, Predicted so far 5.98 kWh with 0.0 kWh car excluded and 1.42 kWh import ignored and 0 forecast extra, Actual so far 6.47 kWh with 0.0 kWh car excluded and 0.53 kWh import ignored
2024-05-20 13:16:40.471748 INFO pred_bat: Inverter 0: Type GE GivEnergy
2024-05-20 13:16:40.472349 INFO pred_bat: Inverter 0 using Rest API http://homeassistant.local:6345
2024-05-20 13:16:41.503836 INFO pred_bat: Inverter 0 does not have timed pause support enabled
2024-05-20 13:16:41.508363 INFO pred_bat: Invertor time 2024-05-20 13:16:00+01:00 AppDaemon time 2024-05-20 13:16:41.503955+01:00 difference -0.7 minutes
2024-05-20 13:16:41.509212 INFO pred_bat: Reserve min: 5.0% Battery_min:4.0%
2024-05-20 13:16:41.512142 INFO pred_bat: Inverter 0 with soc_max 9.52 kWh nominal_capacity 9.52 kWh battery rate raw 3600 w charge rate 3.6 kW discharge rate 3.6 kW battery_rate_min 0.0 w ac limit 5.0 kW export limit 5.0 kW reserve 5.0 % current_reserve 5.0 %
2024-05-20 13:16:43.747189 INFO pred_bat: Inverter 0 SOC: 9.14 kW 96 % Current charge rate 380.0 w Current discharge rate 3600.0 w Current power -390.0 w Current voltage 52.0
2024-05-20 13:16:43.747884 INFO pred_bat: Inverter 0 scheduled charge enable is True
2024-05-20 13:16:43.748547 INFO pred_bat: Inverter 0 charge windows currently [{'start': 300, 'end': 960, 'average': 0}, {'start': 1740, 'end': 2400, 'average': 0}]
2024-05-20 13:16:43.749249 INFO pred_bat: Inverter 0 Charge settings: 05-20 05:00:00-05-20 16:00:00 limit 97.0 power 0.38 kW
2024-05-20 13:16:43.750434 INFO pred_bat: Inverter 0 Wrote scheduled_discharge_enable to False successfully and got off
2024-05-20 13:16:43.751108 INFO pred_bat: Inverter 0 sensor.predbat_GE_0_scheduled_discharge_enable set to False
2024-05-20 13:16:43.751819 INFO pred_bat: Inverter 0 scheduled discharge enable is False
2024-05-20 13:16:43.752539 INFO pred_bat: Inverter 0 discharge windows currently [{'start': 0, 'end': 0, 'average': 0}]
2024-05-20 13:16:43.752971 INFO pred_bat: Find charge curve with sensors sensor.givtcp_fd2311g523_soc_kwh and number.givtcp_fd2311g523_battery_charge_rate and predbat.status and sensor.givtcp_fd2311g523_battery_power
2024-05-20 13:16:47.035204 INFO pred_bat: Find charge curve has 15.0 days of data, max days 15
2024-05-20 13:16:47.046382 INFO pred_bat: Curve Percent: 99-99 at 05-06 04:48:00 took 2 minutes charged 0.1 curve 0.14 average_power 506.0
2024-05-20 13:16:47.048503 INFO pred_bat: Curve Percent: 98-98 at 05-06 04:46:00 took 7 minutes charged 0.1 curve 0.24 average_power 838.0
2024-05-20 13:16:47.049873 INFO pred_bat: Curve Percent: 97-97 at 05-14 04:57:00 took 5 minutes charged 0.1 curve 0.24 average_power 838.0
2024-05-20 13:16:47.050571 INFO pred_bat: Curve Percent: 97-97 at 05-13 04:55:00 took 5 minutes charged 0.1 curve 0.24 average_power 838.0
2024-05-20 13:16:47.051904 INFO pred_bat: Curve Percent: 97-97 at 05-06 04:39:00 took 6 minutes charged 0.1 curve 0.33 average_power 1169.0
2024-05-20 13:16:47.053149 INFO pred_bat: Curve Percent: 96-96 at 05-14 04:52:00 took 5 minutes charged 0.09 curve 0.34 average_power 1170.0
2024-05-20 13:16:47.053906 INFO pred_bat: Curve Percent: 96-96 at 05-13 04:50:00 took 5 minutes charged 0.09 curve 0.33 average_power 1169.0
2024-05-20 13:16:47.055257 INFO pred_bat: Curve Percent: 96-96 at 05-06 04:33:00 took 4 minutes charged 0.09 curve 0.43 average_power 1501.0
2024-05-20 13:16:47.056400 INFO pred_bat: Curve Percent: 95-95 at 05-15 05:00:00 took 3 minutes charged 0.1 curve 0.52 average_power 1833.0
2024-05-20 13:16:47.057063 INFO pred_bat: Curve Percent: 95-95 at 05-14 04:47:00 took 3 minutes charged 0.1 curve 0.43 average_power 1501.0
2024-05-20 13:16:47.057743 INFO pred_bat: Curve Percent: 95-95 at 05-13 04:45:00 took 3 minutes charged 0.1 curve 0.52 average_power 1833.0
2024-05-20 13:16:47.059052 INFO pred_bat: Curve Percent: 95-95 at 05-06 04:29:00 took 3 minutes charged 0.1 curve 0.43 average_power 1501.0
2024-05-20 13:16:47.060287 INFO pred_bat: Curve Percent: 94-94 at 05-15 04:57:00 took 7 minutes charged 0.1 curve 0.52 average_power 1833.0
2024-05-20 13:16:47.060899 INFO pred_bat: Curve Percent: 94-94 at 05-15 04:52:00 took 2 minutes charged 0.1 curve 0.0 average_power 4.0
2024-05-20 13:16:47.061573 INFO pred_bat: Curve Percent: 94-94 at 05-14 04:44:00 took 3 minutes charged 0.1 curve 0.52 average_power 1832.0
2024-05-20 13:16:47.062280 INFO pred_bat: Curve Percent: 94-94 at 05-13 04:42:00 took 3 minutes charged 0.1 curve 0.52 average_power 1833.0
2024-05-20 13:16:47.063601 INFO pred_bat: Curve Percent: 94-94 at 05-06 04:26:00 took 3 minutes charged 0.1 curve 0.62 average_power 2165.0
2024-05-20 13:16:47.064830 INFO pred_bat: Curve Percent: 93-93 at 05-15 04:50:00 took 3 minutes charged 0.1 curve 0.62 average_power 2164.0
2024-05-20 13:16:47.065613 INFO pred_bat: Curve Percent: 93-92 at 05-14 04:41:00 took 3 minutes charged 0.19 curve 0.62 average_power 2164.0
2024-05-20 13:16:47.066331 INFO pred_bat: Curve Percent: 93-92 at 05-13 04:39:00 took 3 minutes charged 0.19 curve 0.62 average_power 2165.0
2024-05-20 13:16:47.067635 INFO pred_bat: Curve Percent: 93-93 at 05-06 04:23:00 took 2 minutes charged 0.1 curve 0.71 average_power 2496.0
2024-05-20 13:16:47.068866 INFO pred_bat: Curve Percent: 92-92 at 05-15 04:47:00 took 2 minutes charged 0.09 curve 0.71 average_power 2496.0
2024-05-20 13:16:47.070458 INFO pred_bat: Curve Percent: 92-92 at 05-06 04:21:00 took 3 minutes charged 0.09 curve 0.81 average_power 2828.0
2024-05-20 13:16:47.072448 INFO pred_bat: Curve Percent: 91-91 at 05-06 04:18:00 took 2 minutes charged 0.1 curve 0.81 average_power 2828.0
2024-05-20 13:16:47.073751 INFO pred_bat: Curve Percent: 90-90 at 05-14 04:36:00 took 2 minutes charged 0.09 curve 1.0 average_power 3492.0
2024-05-20 13:16:47.076113 INFO pred_bat: Curve Percent: 89-89 at 05-14 04:34:00 took 1 minutes charged 0.1 curve 1.0 average_power 3492.0
2024-05-20 13:16:47.078202 INFO pred_bat: Curve Percent: 88-88 at 05-14 04:33:00 took 2 minutes charged 0.1 curve 1.0 average_power 3492.0
2024-05-20 13:16:47.080323 INFO pred_bat: Curve Percent: 87-87 at 05-14 04:31:00 took 2 minutes charged 0.1 curve 1.0 average_power 3491.0
2024-05-20 13:16:47.082492 INFO pred_bat: Curve Percent: 86-86 at 05-14 04:29:00 took 2 minutes charged 0.09 curve 1.0 average_power 3491.0
2024-05-20 13:16:47.083995 INFO pred_bat: Curve before adjustment is: {99: 0.14, 98: 0.24, 97: 0.27, 96: 0.37, 95: 0.47, 94: 0.44, 93: 0.64, 92: 0.69, 91: 0.81, 90: 1.0, 89: 1.0, 88: 1.0, 87: 1.0, 86: 1.0}
2024-05-20 13:16:47.084676 INFO pred_bat: Curve curve can be entered into apps.yaml or set to auto:
  battery_charge_power_curve:
    100 : 0.14
    99 : 0.14
    98 : 0.24
    97 : 0.27
    96 : 0.37
    95 : 0.47
    94 : 0.44
    93 : 0.64
    92 : 0.69
    91 : 0.81

2024-05-20 13:16:47.085150 INFO pred_bat: Find discharge curve with sensors sensor.givtcp_fd2311g523_soc_kwh and number.givtcp_fd2311g523_battery_discharge_rate and predbat.status and sensor.givtcp_fd2311g523_battery_power
2024-05-20 13:16:50.101411 INFO pred_bat: Find discharge curve has 15.0 days of data, max days 15
2024-05-20 13:16:50.133677 INFO pred_bat: Note: Can not find battery charge curve (no final curve), one of the required settings for predbat_status, soc_kw, battery_power and charge_rate do not have history, check apps.yaml
2024-05-20 13:16:50.134245 INFO pred_bat: Found 1 inverters totals: min reserve 0.48 current reserve 0.48 soc_max 9.52 soc 9.14 charge rate 0.38 kW discharge rate 3.5999999999999996 kW battery_rate_min 0.0 w ac limit 5.0 export limit 5.0 kW loss charge 3 % loss discharge 3 % inverter loss 4 %
2024-05-20 13:16:50.183584 INFO pred_bat: Base charge    window [ 20-05 05:00:00 - 20-05 16:00:00 @ 0p 97%, 21-05 05:00:00 - 21-05 16:00:00 @ 0p 97% ]
2024-05-20 13:16:50.184154 INFO pred_bat: Base discharge window [ 20-05 00:00:00 - 20-05 00:00:00 @ 0p 100.0% ]
2024-05-20 13:16:50.184502 INFO pred_bat: Will recompute the plan as it is invalid
2024-05-20 13:16:50.185091 INFO pred_bat: Best charge    window [ 20-05 05:00:00 - 20-05 16:00:00 @ 24.21p 97.0%, 20-05 19:00:00 - 21-05 02:00:00 @ 24.21p 97.0%, 21-05 02:00:00 - 21-05 05:00:00 @ 14.52p 97.0%, 21-05 05:00:00 - 21-05 16:00:00 @ 24.21p 97.0% ]
2024-05-20 13:16:50.185991 INFO pred_bat: Best discharge window [ 20-05 13:00:00 - 20-05 13:30:00 @ 14.84p 100.0%, 20-05 13:30:00 - 20-05 14:00:00 @ 14.84p 100.0%, 20-05 14:00:00 - 20-05 14:30:00 @ 14.84p 100.0%, 20-05 14:30:00 - 20-05 15:00:00 @ 14.84p 100.0%, 20-05 15:00:00 - 20-05 15:30:00 @ 14.84p 100.0%, 20-05 15:30:00 - 20-05 16:00:00 @ 14.84p 100.0%, 20-05 16:00:00 - 20-05 16:30:00 @ 24.52p 100.0%, 20-05 16:30:00 - 20-05 17:00:00 @ 24.52p 100.0%, 20-05 17:00:00 - 20-05 17:30:00 @ 24.52p 100.0%, 20-05 17:30:00 - 20-05 18:00:00 @ 24.52p 100.0%, 20-05 18:00:00 - 20-05 18:30:00 @ 24.52p 100.0%, 20-05 18:30:00 - 20-05 19:00:00 @ 24.52p 100.0%, 20-05 19:00:00 - 20-05 19:30:00 @ 14.84p 100.0%, 20-05 19:30:00 - 20-05 20:00:00 @ 14.84p 100.0%, 20-05 20:00:00 - 20-05 20:30:00 @ 14.84p 100.0%, 20-05 20:30:00 - 20-05 21:00:00 @ 14.84p 100.0%, 20-05 21:00:00 - 20-05 21:30:00 @ 14.84p 100.0%, 20-05 21:30:00 - 20-05 22:00:00 @ 14.84p 100.0%, 20-05 22:00:00 - 20-05 22:30:00 @ 14.84p 100.0%, 20-05 22:30:00 - 20-05 23:00:00 @ 14.84p 100.0%, 20-05 23:00:00 - 20-05 23:30:00 @ 14.84p 100.0%, 20-05 23:30:00 - 21-05 00:00:00 @ 14.84p 100.0%, 21-05 00:00:00 - 21-05 00:30:00 @ 14.84p 100.0%, 21-05 00:30:00 - 21-05 01:00:00 @ 14.84p 100.0%, 21-05 01:00:00 - 21-05 01:30:00 @ 14.84p 100.0%, 21-05 01:30:00 - 21-05 02:00:00 @ 14.84p 100.0%, 21-05 05:00:00 - 21-05 05:30:00 @ 14.84p 100.0%, 21-05 05:30:00 - 21-05 06:00:00 @ 14.84p 100.0%, 21-05 06:00:00 - 21-05 06:30:00 @ 14.84p 100.0%, 21-05 06:30:00 - 21-05 07:00:00 @ 14.84p 100.0%, 21-05 07:00:00 - 21-05 07:30:00 @ 14.84p 100.0%, 21-05 07:30:00 - 21-05 08:00:00 @ 14.84p 100.0%, 21-05 08:00:00 - 21-05 08:30:00 @ 14.84p 100.0%, 21-05 08:30:00 - 21-05 09:00:00 @ 14.84p 100.0%, 21-05 09:00:00 - 21-05 09:30:00 @ 14.84p 100.0%, 21-05 09:30:00 - 21-05 10:00:00 @ 14.84p 100.0%, 21-05 10:00:00 - 21-05 10:30:00 @ 14.84p 100.0%, 21-05 10:30:00 - 21-05 11:00:00 @ 14.84p 100.0%, 21-05 11:00:00 - 21-05 11:30:00 @ 14.84p 100.0%, 21-05 11:30:00 - 21-05 12:00:00 @ 14.84p 100.0%, 21-05 12:00:00 - 21-05 12:30:00 @ 14.84p 100.0%, 21-05 12:30:00 - 21-05 13:00:00 @ 14.84p 100.0%, 21-05 13:00:00 - 21-05 13:30:00 @ 14.84p 100.0% ]
2024-05-20 13:16:50.186588 INFO pred_bat: PV Forecast 0.0 kWh and 10% Forecast 0.0 kWh pv cloud factor None
2024-05-20 13:16:50.187916 INFO pred_bat: Load divergence over 4.0 hours mean 697.36 W, min 250.2 W, max 1249.8 W, std dev 251.41 W, divergence 36.05%
2024-05-20 13:16:50.225492 INFO pred_bat: Yesterday basic charge window best: [{'start': 120, 'end': 300, 'average': 14.52}] charge limit best: [9.52]
2024-05-20 13:16:50.270207 INFO pred_bat: ERROR: Exception raised 1470

**Comment by RobinCu** [comment]
Can you try the fix here: https://github.com/springfall2008/batpred/issues/1107

**Comment by Neomancer86** [comment]
Confirmed, https://github.com/springfall2008/batpred/issues/1107#issuecomment-2120544185 fixed this for me.

I tried to update from the GUI but updated in apps.yaml resolved it.

**Comment by springfall2008** [comment]
Fixed in latest release v7.19.6
